### PR TITLE
Enhancements

### PR DESCRIPTION
**Pull request by kayladnls**
- [x] Adds Resource Factory 
- [x] Allows Error Collection to be instantiated with an array

Blocks: https://github.com/refinery29/zeppelin/pull/88

### CODE REVIEW

File: .php_cs.cache [hunk 1]
```diff
@@ -1 +1 @@
...
...i:3177477144;s:45:"spec/Serializer/Error/ErrorCollectionSpec.php";i:160963317;s:35:"spec/Serializer/Error/ErrorSpec.php";i:3714712787;s:43:"spec/Serializer/Link/LinkCollectionSpec.php";i:2879821438;s:45:"spec/Serializer/Pagination/PaginationSpec.php";i:3629134900;s:30:"spec/Serializer/ResultSpec.php";i:3891965103;s:28:"src/Resource/Error/Error.php";i:3684027032;s:38:"src/Resource/Error/ErrorCollection.php";i:2894553678;s:30:"src/Resource/Link/HasLinks.php";i:2832874921;s:26:"src/Resource/Link/Link.php";i:2544502352;s:36:"src/Resource/Link/LinkCollection.php";i:870395955;s:38:"src/Resource/Pagination/Pagination.php";i:1815957060;s:32:"src/Resource/ResourceFactory.php";i:4114760010;s:23:"src/Resource/Result.php";i:3802206450;s:20:"src/ResponseBody.php";i:2198644468;s:30:"src/Serializer/Error/Error.php";i:1523036371;s:40:"src/Serializer/Error/ErrorCollection.php";i:2008055125;s:32:"src/Serializer/HasSerializer.php";i:2413301796;s:38:"src/Serializer/Link/LinkCollection.php";i:2170125106;s:40:"src/Serializer/Pagination/Pagination.php";i:2232211960;s:25:"src/Serializer/Result.php";i:3627550639;s:29:"src/Serializer/Serializer.php";i:1692954048;s:24:"src/TopLevelResource.php";i:17425208;}}
```
Comment (localheinz) [inline .php_cs.cache]: @kayladnls 

Not sure, is it intended to check in `.php_cs.cache`?

File: src/Resource/Error/ErrorCollection.php [hunk 2]
```diff
@@ -8,6 +8,17 @@ class ErrorCollection implements HasSerializer
 {
     private $errors = [];
 
+    public function __construct(array $errors = [])
+    {
+        foreach ($errors as $error) {
```
Comment (localheinz) [inline src/Resource/Error/ErrorCollection.php]: @kayladnls 

:wink: Reminds me of this: http://ocramius.github.io/extremely-defensive-php/#/75!

Comment (localheinz) [inline src/Resource/Error/ErrorCollection.php]: @kayladnls 

How about using [`beberlei/assert`](https://github.com/beberlei/assert) here?

```php
use Assert\Assertion;

public function __construct(array $errors = []) 
{
    Assertion::allIsInstanceOf($errors, Error::class, sprintf(
        'Each element of parameter "errors" needs to be an instance of %s",
        Error::class
    ));

    $this->errors = $errors;
}

File: src/Resource/Error/ErrorCollection.php [hunk 3]
```diff
@@ -8,6 +8,17 @@ class ErrorCollection implements HasSerializer
 {
     private $errors = [];
 
+    public function __construct(array $errors = [])
```
Comment (localheinz) [inline src/Resource/Error/ErrorCollection.php]: @kayladnls 

Can we add a docblock here?

```php
/**
 * @param Error[] $errors
 */
```

File: src/Resource/Error/ErrorCollection.php (hunk 2 above)
Comment (kayladnls) [inline src/Resource/Error/ErrorCollection.php]: Assert isn't a dep of this repo. Is it worth making it one for this purpose?

File: src/Resource/ResourceFactory.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    /**
+     * @return ErrorCollection
+     */
+    public static function errorCollection()
+    {
+        return new ErrorCollection();
+    }
+
+    /**
+     * @param $title
```
Comment (localheinz) [inline src/Resource/ResourceFactory.php]: @kayladnls 

Can we add types here, too?

File: src/Resource/Error/ErrorCollection.php (hunk 2 above)
Comment (localheinz) [inline src/Resource/Error/ErrorCollection.php]: @kayladnls 

If we want to guard against invalid inputs in other places, too, I'd highly recommend it!

File: src/Resource/ResourceFactory.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+namespace Refinery29\ApiOutput\Resource;
+
+use Refinery29\ApiOutput\Resource\Error\Error;
+use Refinery29\ApiOutput\Resource\Error\ErrorCollection;
+use Refinery29\ApiOutput\Resource\Link\LinkCollection;
+
+abstract class ResourceFactory
+{
+    /**
+     * @param $data
```
Comment (localheinz) [inline src/Resource/ResourceFactory.php]: @kayladnls 

Can we add the type here, too?

**Comment by localheinz** [comment]
@kayladnls 

:+1: Other than the small comments, this is good to merge!

**Pull request merged**